File: libraries/NTL/Combination.cpp

ll comb[MAX_N+1][MAX_N+1];
for (int i = 0; i <= MAX_N; i++) {
    for (int j = 0; j <= i; j++) {
        if (j == 0 || j == i)
            comb[i][j] = 1;
        else
            comb[i][j] = comb[i - 1][j - 1] + comb[i - 1][j];
    }
}

＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝


const ll MAX_N = 3e6 + 10;
const ll MOD = 1e9+7;
ll Inv[MAX_N];
ll fact[MAX_N];
ll factInv[MAX_N];
void Inv_init(){
    Inv[1] = 1;
    for(int i = 2; i < MAX_N; i++) {
        Inv[i] = Inv[MOD%i] * (MOD - MOD/i) % MOD;
    }
}
void fact_init(){
    fact[0] = fact[1] = factInv[0] = factInv[1] = 1;
    for(int i = 2; i < MAX_N; i++){
        fact[i] = (fact[i-1] * i) % MOD;
        factInv[i] = (factInv[i-1] * Inv[i])%MOD;
    }
}

// a^b % MOD;
ll powmod(ll a,ll b) {ll res=1;a%=MOD; for(;b;b>>=1){if(b&1)res=res*a%MOD;a=a*a%MOD;}return res;}
// 逆元 1/a % MOD
ll inversemod(ll a) { return powmod(a,MOD - 2);}

// 1 以上 N 以下の N 個の整数の中から，相異なる K 個の整数を選ぶパターンの数
ll nCr(int n,int r) {
    if(n < r) return 0;
    return (fact[n] * (factInv[r] * factInv[n-r] % MOD)) % MOD;
}
// 1 以上 N 以下の N 個の整数の中から，相異なる K 個の整数を選び，順番に並べるパターンの数
ll nPr(int n,int r){
    if(n < r) return 0;
    return (fact[n]*factInv[n-r])%MOD;
}
// 1 以上 N 以下の N 個の整数の中から，重複を許して K 個の整数を選ぶパターンの数
ll nHr(int n,int r){
    if(n == 0 && r == 0) return 1;
    return nCr(n+r-1,r);
}

void init(){
    Inv_init();
    fact_init();
}
============================================================
C(n,m) = C(n,m-1) * (n-m+1) / m.
C(n,m) = C(n,n-m)

ll C(int n,int m){
    m = m<(n-m) ? m : (n-m);
    if(m == 0) return 1;
    ll ans = 1;
    for(int i=1;i<=m;i++){
        ans = ans *(n-i+1)/i;
    }
    return ans;
}

File: for_rust/tools/combination.rs
static MOD: i64 = 1_000_000_007;
fn fact_init(MAX_N: usize) -> (Vec<i64>, Vec<i64>, Vec<i64>) {
    let mut Inv: Vec<i64> = vec![0; MAX_N];
    let mut fact: Vec<i64> = vec![0; MAX_N];
    let mut factInv: Vec<i64> = vec![0; MAX_N];

    Inv[1] = 1;
    for i in 2..MAX_N as i64 {
        Inv[i as usize] = Inv[(MOD % i) as usize] * (MOD - MOD / i) % MOD;
    }

    fact[0] = 1;
    fact[1] = 1;
    factInv[0] = 1;
    factInv[1] = 1;
    for i in 2..MAX_N as i64 {
        fact[i as usize] = (fact[(i - 1) as usize] * i) % MOD;
        factInv[i as usize] = (factInv[(i - 1) as usize] * Inv[i as usize]) % MOD;
    }
    (Inv, fact, factInv)

    //let (mut Inv,mut fact,mut factInv) = fact_init((N+1) as usize);
}

fn powmod(a: i64, b: i64,_MOD:i64) -> i64 {
    let mut a: i64 = a;
    let mut b: i64 = b;
    let mut res: i64 = 1;
    a %= _MOD;
    while b != 0 {
        if b % 2 == 1 {
            res = res * a % _MOD;
        }
        a = a * a % _MOD;
        b >>= 1;
    }
    res
}

fn inversemod(a: i64, _MOD: i64) -> i64 {
    powmod(a, _MOD - 2,_MOD)
}

fn nCr(n: usize, r: usize,_MOD:i64,fact:&Vec<i64>,factInv:&Vec<i64>) -> i64 {
    let mut ret :i64 = 0;
    if n >= r { ret = fact[n]*factInv[r]%_MOD*factInv[n-r]%_MOD; }
    ret
}